src/evict/evict_lru.c

```diff
@@ -843,6 +835,164 @@ __wt_evict_file_exclusive_off(WT_SESSION_IMPL *session)
+ * see from the creation of another eviction worker to justify the creation of
+ * another one or to justify keeping around the previously created thread.
+ */
+#define	EVICT_TUNE_THRESHOLD 5 /* Percent eviction rate */
+
+/*
+ * __evict_tune_workers --
+ *      Decide if we want to increase or decrease the number of active eviction
+ *      workers. This function needs to be called relatively frequently and
+ *      must be performed by a thread with the smallest id in the eviction
+ *      thread group.
+ */
+static int
+__evict_tune_workers(WT_SESSION_IMPL *session)
+{
+	WT_CACHE *cache;
+	WT_CONNECTION_IMPL *conn;
+	WT_DECL_RET;
+	struct timespec current_time;
+	int pct_diff;
+	uint64_t cur_threads, delta_millis, delta_pages, i, target_threads;
+	uint64_t pgs_evicted_cur, pgs_evicted_persec_cur;
+	bool try_create_threads, try_remove_threads;
+
+	conn = S2C(session);
+	cache = conn->cache;
+	try_create_threads = try_remove_threads = false;
+
+	WT_ASSERT(session, conn->evict_threads.threads[0]->session == session);
+	pgs_evicted_persec_cur = 0;
+
+	WT_RET(__wt_epoch(session, &current_time));
+
+	/*
+	 * Every EVICT_TUNE_PERIOD seconds record the number of
+	 * pages evicted per second observed in the previous period.
+	 */
+	if (WT_TIMEDIFF_SEC(
+	    current_time, conn->evict_tune_last_time) < EVICT_TUNE_PERIOD)
+		return (0);
+
+	pgs_evicted_cur = cache->pages_evict;
+
+	/*
+	 * If we have recorded the number of pages evicted at the end of
+	 * the previous measurement interval, we can compute the eviction
+	 * rate in evicted pages per second achieved during the current
+	 * measurement interval.
+	 * Otherwise, we just record the number of evicted pages and return.
+	 */
+	if (conn->evict_tune_pgs_last == 0)
+		goto err;
+
+	delta_millis = WT_TIMEDIFF_MS(current_time, conn->evict_tune_last_time);
+	delta_pages = pgs_evicted_cur - conn->evict_tune_pgs_last;
+	pgs_evicted_persec_cur = (delta_pages * WT_THOUSAND) / delta_millis;
+
+	/*
+	 * If we have not recorded the pages evicted per second rate in the
+	 * previous interval, we need to wait for another measurement interval
+	 * until we can record that rate, because we cannot make any tuning
+	 * decisions without it.
+	 */
+	if (conn->evict_tune_pg_sec_last == 0)
+		goto err;
+
+	pct_diff = (int)(pgs_evicted_persec_cur - conn->evict_tune_pg_sec_last)
+	    * 100 / (int) conn->evict_tune_pg_sec_last;
+
+	/*
+	 * If we previously added threads and saw no improvement, we will remove
+	 * half of the threads currently working.
+	 * If we added threads and saw improvement, we will keep adding.
+	 * If we previously removed threads and so saw no degradation, we will
+	 * keep removing.
+	 * If we previously removed threads and saw degradation, we will add.
+	 */
+	if (conn->evict_tune_last_action == EVICT_ADDED)
+		if (pct_diff > EVICT_TUNE_THRESHOLD)
+			try_create_threads = true;
+		else
+			try_remove_threads = true;
+	else if (conn->evict_tune_last_action == EVICT_REMOVED)
```

Comment:
@sulabhM: yes, we will keep adding/removing threads to avoid getting stuck in a local minima and to adapt to the dynamically changing workload.